### PR TITLE
Uplift `clippy::cast_ref_to_mut` lint

### PR DESCRIPTION
**Pull request by Urgau**
This PR aims at uplifting the `clippy::cast_ref_to_mut` lint into rustc.

## `cast_ref_to_mut`

> :information_source: The lint has been renamed to `invalid_reference_casting`  in https://github.com/rust-lang/rust/pull/113422 and https://github.com/rust-lang/rust/pull/112431

(deny-by-default)

The `cast_ref_to_mut` lint checks for casts of `&T` to `&mut T` without using interior mutability.

### Example

```rust,compile_fail
fn x(r: &i32) {
    unsafe {
        *(r as *const i32 as *mut i32) += 1;
    }
}
```

### Explanation

Casting `&T` to `&mut T` without interior mutability is undefined behavior, as it's a violation of Rust reference aliasing requirements.

-----

Mostly followed the instructions for uplifting a clippy lint described here: https://github.com/rust-lang/rust/pull/99696#pullrequestreview-1134072751

@rustbot label: +I-lang-nominated
r? compiler

-----

For Clippy:

changelog: Moves: Uplifted `clippy::cast_ref_to_mut` into rustc

### CODE REVIEW

File: compiler/rustc_lint/src/cast_ref_to_mut.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+use rustc_ast::Mutability;
+use rustc_hir::{Expr, ExprKind, MutTy, TyKind, UnOp};
+use rustc_middle::ty;
+
+use crate::{lints::CastRefToMutDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `cast_ref_to_mut` lint checks for casts of `&T` to `&mut T`
+    /// without inferior mutability.
```
Comment (bugadani) [inline compiler/rustc_lint/src/cast_ref_to_mut.rs:9]: Is "inferior" a typo here, or some term I don't understand in this context?

Comment (Urgau) [reply compiler/rustc_lint/src/cast_ref_to_mut.rs]: Just a typo. ;-)

File: compiler/rustc_lint/src/cast_ref_to_mut.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /// ### Explanation
+    ///
+    /// Casting `&T` to `&mut T` without interior mutability is undefined behavior,
+    /// as it's a violation of Rust reference aliasing requirements.
+    ///
+    /// `UnsafeCell` is the only way to obtain aliasable data that is considered
+    /// mutable.
+    CAST_REF_TO_MUT,
+    Deny,
+    "casts of `&T` to `&mut T` without inferior mutability"
```
Comment (bugadani) [inline compiler/rustc_lint/src/cast_ref_to_mut.rs:32]:
```suggestion
    "casts of `&T` to `&mut T` without interior mutability"
```

Here as well :)

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by scottmcm** [comment]
Sadly this wouldn't catch the <https://blog.rust-lang.org/2017/02/09/Rust-1.15.1.html#whats-in-1151-stable> issue, but still sounds like a good lint.

It would be nice if it could also catch things like
```rust=
*ptr::from_ref(r).cast_mut() += 1;
```
but that doesn't need to block accepting this.

@rfcbot reviewed

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Pull request merged**